docparse/find.go

```diff
@@ -172,10 +179,243 @@ func FindType(currentFile, pkgPath, name string) (
+
+	// Scan all fields of f if it refers to a struct. Do this after storing the
+	// reference in prog.References to prevent cyclic lookup issues.
+	for _, f := range st.Fields.List {
+		if goutil.TagName(f, "json") == "-" {
+			continue
+		}
+
+		err := findNested(prog, context, f, foundPath, pkg)
+		if err != nil {
+			return nil, fmt.Errorf("\n  findNested: %v", err)
+		}
+	}
+
+	// Convert to JSON Schema.
+	schema, err := structToSchema(prog, name, ref)
+	if err != nil {
+		return nil, fmt.Errorf("%v can not be converted to JSON schema: %v", name, err)
+	}
+	ref.Schema = schema
+	prog.References[ref.Lookup] = ref
+
+	return &ref, nil
+}
+
+func findNested(prog *Program, context string, f *ast.Field, filePath, pkg string) error {
+	var name *ast.Ident
+
+	sw := f.Type
+start:
+	switch typ := sw.(type) {
+
+	// Pointer type; we don't really care about this for now, so just read over
+	// it.
+	case *ast.StarExpr:
+		sw = typ.X
+		goto start
+
+	// Simple identifiers such as "string", "int", "MyType", etc.
+	case *ast.Ident:
+		if !builtInType(typ.Name) {
+			name = typ
+		}
+
+	// An expression followed by a selector, e.g. "pkg.foo"
+	case *ast.SelectorExpr:
+		pkgSel, ok := typ.X.(*ast.Ident)
+		if !ok {
+			return fmt.Errorf("typ.X is not ast.Ident: %#v", typ.X)
+		}
+		name = typ.Sel
+		pkg = pkgSel.Name
+
+	// Array and slices.
+	case *ast.ArrayType:
+		asw := typ.Elt
+
+	arrayStart:
+		switch elementType := asw.(type) {
+
+		// Ignore *
+		case *ast.StarExpr:
+			asw = elementType.X
+			goto arrayStart
+
+		// Simple identifier
+		case *ast.Ident:
+			if !builtInType(elementType.Name) {
+				name = elementType
+			}
+
+		// "pkg.foo"
+		case *ast.SelectorExpr:
+			pkgSel, ok := elementType.X.(*ast.Ident)
+			if !ok {
+				return fmt.Errorf("elementType.X is not ast.Ident: %#v",
+					elementType.X)
+			}
+			name = elementType.Sel
+			pkg = pkgSel.Name
+		}
+	}
+
+	if name == nil {
+		return nil
+	}
+
+	lookup := pkg + "." + name.Name
+	if i := strings.LastIndex(pkg, "/"); i > -1 {
+		lookup = pkg[i+1:] + "." + name.Name
+	}
+
+	// Don't need to add stuff we map to Go primitives.
+	if _, ok := mapTypes[lookup]; ok {
+		return nil
+	}
+
+	if _, ok := prog.References[lookup]; !ok {
+		err := resolveType(prog, context, name, filePath, pkg)
+		if err != nil {
+			return fmt.Errorf("%v.%v: %v", pkg, name, err)
+		}
+	}
+	return nil
+}
+
+func resolveType(prog *Program, context string, typ *ast.Ident, filePath, pkg string) error {
+	var ts *ast.TypeSpec
+	if typ.Obj == nil {
+		var err error
+		ts, _, _, err = findType(filePath, pkg, typ.Name)
+		if err != nil {
+			return err
+		}
+	} else {
+		var ok bool
+		ts, ok = typ.Obj.Decl.(*ast.TypeSpec)
+		if !ok {
+			return fmt.Errorf("resolveType: not a type declaration but %T",
+				typ.Obj.Decl)
+		}
+	}
+
+	// Only need to add struct types. Stuff like "type Foo string" gets added as
+	// simply a string.
+	_, ok := ts.Type.(*ast.StructType)
+	if !ok {
+		return nil
```

Comment:
I was wrong about the issue with custom slice types being fixed, it will not error anymore but the details for its type will never end up in `prog.References`, it ends up returning here.

Should it check here if its an ArrayType then resolve its `Elt` field?

Reply:
It shouldn't use References at all for custom slice and map types; there is no real point in it. It should "resolve" it to the "real" type, just like it already does when you refer `type foo string`. I created a WIP PR at #15.